File: samples/client/3_0_3_unit_test/java/src/main/java/org/openapijsonschematools/client/schemas/validation/PropertyNamesValidator.java
package org.openapijsonschematools.client.schemas.validation;

import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class PropertyNamesValidator implements KeywordValidator {
    public final Class<? extends JsonSchema> propertyNames;

    public PropertyNamesValidator(Class<? extends JsonSchema> propertyNames) {
        this.propertyNames = propertyNames;
    }

    @Override
    public @Nullable PathToSchemasMap validate(
        JsonSchema schema,
        @Nullable Object arg,
        ValidationMetadata validationMetadata,
        @Nullable List<PathToSchemasMap> containsPathToSchemas
    ) {
        if (!(arg instanceof Map)) {
            return null;
        }
        JsonSchema propertyNamesSchema = JsonSchemaFactory.getInstance(propertyNames);
        for (Object objKey: ((Map<?, ?>) arg).keySet()) {
            if (objKey instanceof String key) {
                List<Object> propPathToItem = new ArrayList<>(validationMetadata.pathToItem());
                propPathToItem.add(key);
                ValidationMetadata keyValidationMetadata = new ValidationMetadata(
                        propPathToItem,
                        validationMetadata.configuration(),
                        validationMetadata.validatedPathToSchemas(),
                        validationMetadata.seenClasses()
                );
                JsonSchema.validate(propertyNamesSchema, key, keyValidationMetadata);
            }
        }
        return null;
    }
}

File: samples/client/3_0_3_unit_test/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/PropertyNamesValidator.kt
package org.openapijsonschematools.client.schemas.validation

import org.openapijsonschematools.client.exceptions.ValidationException

class PropertyNamesValidator : KeywordValidator {
    @Throws(ValidationException::class)
    override fun validate(
        data: ValidationData
    ): PathToSchemasMap? {
        val propertyNames: Class<out JsonSchema<*>> = data.schema.propertyNames ?: return null
        if (data.arg !is Map<*, *>) {
            return null
        }
        val propertyNamesSchema = JsonSchemaFactory.getInstance(propertyNames)
        for (objKey in data.arg.keys) {
            if (objKey is String) {
                val propPathToItem: List<Any> = data.validationMetadata.pathToItem + objKey
                val keyValidationMetadata = ValidationMetadata(
                    propPathToItem,
                    data.validationMetadata.configuration,
                    data.validationMetadata.validatedPathToSchemas,
                    data.validationMetadata.seenClasses
                )
                JsonSchema.validate(propertyNamesSchema, objKey, keyValidationMetadata)
            }
        }
        return null
    }
}
